examples/sparse/ba_dense.py

```diff
@@ -0,0 +1,114 @@
+        - camera_index_of_observations: torch.Tensor (n_observations,)
+            contains indices of cameras (from 0 to n_cameras - 1) involved in each observation.
+        - point_index_of_observations: torch.Tensor (n_observations,)
+            contains indices of points (from 0 to n_points - 1) involved in each observation.
+    num_opt_steps : int, optional
+        Number of optimization steps, by default 1000
+
+    Returns
+    -------
+    camera_extrinsics : pp.LieTensor (n_cameras, 7)
+        The optimized camera extrinsics.
+    points_3d : torch.Tensor (n_points, 3)
+        The optimized 3D points.
+    """
+    print(f'Solving {dataset["problem_name"]}')
+    print(f'Number of optimization steps: {num_opt_steps}')
+
+    # use pytorch's adam for dense ba, device is automatically set to cuda if available
+    device = 'cuda' if torch.cuda.is_available() else 'cpu'
+    print(f'Using device: {device}')
+
+    for k, v in dataset.items():
+      if isinstance(v, torch.Tensor):
+          dataset[k] = v.to(device)
+
+    dataset['camera_extrinsics'].requires_grad_(True)
+    dataset['points_3d'].requires_grad_(True)
+
+    optimizer = torch.optim.Adam([dataset['camera_extrinsics'], dataset['points_3d']], lr=3e-4)
+    t = trange(num_opt_steps, desc='Loss', leave=True)
+    loss_history = []
+    loss_vector_history = []
+
+    for i in t:
+        optimizer.zero_grad()
+        loss = reprojerr(dataset['camera_extrinsics'],
+                 dataset['points_3d'],
+                 dataset['points_2d'],
+                 dataset['camera_intrinsics'],
+                 dataset['camera_distortions'],
+                 dataset['point_index_of_observations'],
+                 dataset['camera_index_of_observations'])
+
+        loss_vector_history.append(loss.detach().cpu().numpy())
+        loss = torch.mean(loss)
+        loss.backward()
+        optimizer.step()
+
+        loss_history.append(loss.item())
+        t.set_description(f'Loss: {loss.item():.4f}')
+
+    print(f'Initial loss: {loss_history[0]}')
+    print(f'Final loss: {loss_history[-1]}')
+    visualize_loss_history(loss_history)
+    visualize_loss_per_observation(loss_vector_history[0], path_to_img="loss_per_observation_initial.png", title="Loss per observation (initial)")
+    visualize_loss_per_observation(loss_vector_history[-1], path_to_img="loss_per_observation_final.png", title="Loss per observation (final)")
+
+    return dataset['camera_extrinsics'], dataset['points_3d']
+
+
+if __name__ == '__main__':
+    # parse arguments
+    parser = argparse.ArgumentParser(description='Bundle adjustment for the BAL dataset.')
+    parser.add_argument('--steps', type=int, default=1000, help='Number of optimization steps.')
+    parser.add_argument('--dataset', type=str, default='ladybug', help='BAL dataset name.')
+    parser.add_argument('--problem', type=str, default='problem-49-7776-pre', help='BAL problem name.')
+    args = parser.parse_args()
```

Comment:
I think we can add a `-h` or `--help` argument, which shows an example command line and direct the user to the `https://grail.cs.washington.edu/projects/bal/` webpage for possible values of the `--dataset` and `--problem` arguments.

Reply:
I refactored the help message into
```
usage: ba_dense.py [-h] [--steps STEPS] [--dataset DATASET] [--problem PROBLEM]

This script runs bundle adjustment on the BAL dataset. To understand what datasets and problems are available, please refer to https://grail.cs.washington.edu/projects/bal

options:
  -h, --help         show this help message and exit
  --steps STEPS      Number of optimization steps.
  --dataset DATASET  BAL dataset name, please refer to
                     https://grail.cs.washington.edu/projects/bal for possible
                     values.
  --problem PROBLEM  BAL problem name, please refer to
                     https://grail.cs.washington.edu/projects/bal for possible
                     values.

Example of optimizing `problem-49-7776-pre` in the `ladybug` dataset:
 python3 ba_dense.py --dataset ladybug --problem problem-49-7776-pre --steps 1000
```